qmail-qfilter.1
.TH qmail-qfilter 1
.SH NAME
qmail-qfilter \- front end for qmail-queue that does filtering
.SH SYNOPSIS
.B qmail-qfilter
.I filter
[
.I -- filter ...
]
.SH DESCRIPTION
.B qmail-qfilter
sends the message text through each of the filter commands named on
the command line. 
Each filter is run separately, with standard input opened to the input
email, and standard output opened to a new temporary file that will
become the input to either the next filter, or
.BR qmail-queue .
If the filter does not modify the message it passes unchanged to the
next step.
It also makes the envelope available to each filter as
file descriptor 3.  File descriptor 4 is opened to a new temporary file for the
modified envelope, allowing the filter to modify the envelope or the
message.  If the filter does not modify the envelope, the envelope
remains unchanged for either the next filter or 
.BR qmail-queue .  
This provides compatibility for existing filters that do not know about
the envelope.
.B qmail-qfilter
also opens up file descriptor 5 to a temporary file.  If this file is
empty after all the filters have executed, its contents are read and
used to specify a program to execute in place of
.BR qmail-queue .
Each filter on the command line is separated with
.BR -- .  
.SH "RETURN VALUES"
Returns 51 (out of memory), 53 (write error), or 81 (internal error)
if it can't create the temporary files or has problems executing the
filters.
Returns 91 (bad envelope data) if it can't read or parse the envelope
data.
If a filter returns anything other than 0 or 99,
.B qmail-qfilter
returns its exit code.
If a filter returns 99,
.B qmail-qfilter
returns 0 immediately without running any other filters.
Otherwise returns the exit code of
.BR qmail-queue .
.SH ENVIRONMENT
.B qmail-qfilter
sets
.B QMAILPPID
to the PID of its parent (e.g.,
.BR qmail-smtpd )
so that a filter can log the PID of the session that invoked it.
.P
For compatibility with previous versions,
.B qmail-qfilter
sets
.B QMAILUSER
and
.B QMAILHOST
to the user and host portions of the envelope sender address, and
unsets
.BR QMAILNAME .
It also sets
.B QMAILRCPTS
to the list of envelope recipients, each followed by a newline.
.P
It also sets
.B ENVSIZE
to the size of the envelope,
.B MSGSIZE
to the length of the message, and
.B NUMRCPTS
to the number of recipients. These values are updated before each
filter is run.
.P
If
.B QQF_QMAILQUEUE
is set, its value is used in place of
.BR qmail-queue .
.SH "SEE ALSO"
qmail-queue(8)
.SH NOTES
.B $QMAILQUEUE
is deliberately not used to override
.B qmail-queue
in order to avoid recursive loops with configurations that set
.B $QMAILQUEUE
to invoke
.B qmail-qfilter
itself.
.SH WARNINGS
If you are using
.B qmail-inject -n
as one of the filters, you may want to unset
.BR MAILUSER ,
.BR USER ,
and
.B LOGNAME
by using
.B env -u QMAILNAME -u MAILNAME -u NAME qmail-inject -n
as the command to invoke qmail-inject.
Note that the
.B env
command with some OSes doesn't support the
.B -u
option.
.P
A message with an excessive number of recipients (more than 64K bytes of
recipient data on Linux) will cause execution of the filter programs to
fail, and the message to be rejected.
.P
The same temporary file is reused for file descriptor 5 for each filter.
Make sure each filter writes a trailing ASCII NUL byte following the
program name, as multiple filters could otherwise overwrite the value in
undesirable ways.


NEWS
- Set $QMAILPPID in the executed filters so filters know the grandparent
  PID (ie qmail-smtpd, etc). Thanks Amitai Schleier.
-------------------------------------------------------------------------------
Changes in version 2.1

- Fixed a bug triggered by a filter modifying the envelope data and
  changing its length.  Thanks Dale Woolridge.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 2.0

- Added support for modifying the message envelope as well as the body.

- If either the body or envelope output is empty, the input is reused
  as-is.

- Export the envelope size, the message size, and the number of
  recipients to filters through environment variables $ENVSIZE,
  $MSGSIZE, and $NUMRCPTS respectively.

Development of this version has been sponsored by Saffron Solutions, Inc.
http://www.saffron.net/
-------------------------------------------------------------------------------
Changes in version 1.6 (unreleased)

- Added a work-around for systems that lack unsetenv (notably Solaris).

- Fixed a off-by-one bug in parsing sender addresses without an "@".

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.5

- Fixed a bug in handling the empty sender envelope.
- Added code to handle missing sender domains instead of rejecting them.
- Fixed a bug in handling empty messages.
- Don't close standard error when running filters so they can output log
  messages.
-------------------------------------------------------------------------------
Changes in version 1.4

- Added a new "drop message" exit code -- if the filter exits with code
  99, qmail-qfilter immediately exits successfully without calling
  qmail-queue.
- Some documentation fixups.
-------------------------------------------------------------------------------
Changes in version 1.3

- A new environment variable, "QMAILRCPTS" is set to a newline-seperated
  list of the recipients.  Note that if a large number of recipients are
  encountered (ex more than 64KB worth under Linux), this will cause
  execution of the filters to fail and the message will be rejected.  I
  consider this acceptable, since 64KB worth of recipients is likely
  more than 1000 anyways.
- Removed some GNU-specific constructs from the source and Makefile.
- Included a sample MIME filename extension scanning filter.
-------------------------------------------------------------------------------
Changes in version 1.2

- Fixed a couple of minor bugs in error return codes in run_filters.
-------------------------------------------------------------------------------
Changes in version 1.1

- Fixed a bug in building the linked list of commands to run.
-------------------------------------------------------------------------------


qmail-qfilter.c
/* Copyright (C) 2001,2004-2005 Bruce Guenter <bruceg@em.ca>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <bglibs/sysdeps.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/fcntl.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <signal.h>
#include <unistd.h>

#ifndef TMPDIR
#define TMPDIR "/tmp"
#endif

#ifndef BUFSIZE
#define BUFSIZE 4096
#endif

#ifndef QMAIL_QUEUE
#define QMAIL_QUEUE "/var/qmail/bin/qmail-queue"
#endif

#define QQ_OOM 51
#define QQ_WRITE_ERROR 53
#define QQ_INTERNAL 81
#define QQ_BAD_ENV 91

#define QQ_DROP_MSG 99

#define MSGIN 0
#define MSGOUT 1
#define ENVIN 3
#define ENVOUT 4
#define QQFD 5

static const char* qqargv[2];

/* a replacement for setenv(3) for systems that don't have one */
void mysetenv(const char* key, const char* val, size_t vallen)
{
  char* tmp;
  size_t keylen;
  
  keylen = strlen(key);
  tmp = malloc(keylen + 1 + vallen + 1);
  memcpy(tmp, key, keylen);
  tmp[keylen] = '=';
  memcpy(tmp+keylen+1, val, vallen);
  tmp[keylen+1+vallen] = 0;
  if (putenv(tmp) != 0)
    exit(QQ_OOM);
}

void mysetenvu(const char* key, unsigned long val)
{
  char buf[40];
  int i;
  i = sizeof buf;
  buf[--i] = 0;
  do {
    buf[--i] = (val % 10) + '0';
    val /= 10;
  } while (val > 0);
  mysetenv(key, buf + i, sizeof buf - 1 - i);
}

static size_t env_len = 0;
static size_t msg_len = 0;

/* Parse the sender address into user and host portions */
size_t parse_sender(const char* env)
{
  const char* ptr = env;
  char* at;
  size_t len = strlen(env);
  
  if(*ptr != 'F')
    exit(QQ_BAD_ENV);
  ++ptr;
  
  unsetenv("QMAILNAME");
  
  if(!*ptr) {
    if (putenv("QMAILUSER=") != 0
	|| putenv("QMAILHOST=") != 0)
      exit(QQ_OOM);
    return 2;
  }

  at = strrchr(ptr, '@');
  if(!at) {
    len = strlen(ptr);
    mysetenv("QMAILUSER", ptr, len);
    putenv("QMAILHOST=");
  }
  else {
    len = strlen(at);
    mysetenv("QMAILUSER", ptr, at-ptr);
    mysetenv("QMAILHOST", at+1, len-1);
    ptr = at;
  }
  return ptr + len + 1 - env;
}

void parse_rcpts(const char* env, int offset)
{
  size_t len = env_len - offset;
  const char* ptr = env + offset;
  char* buf = malloc(len);
  char* tmp = buf;
  unsigned long count;
  count = 0;
  while(ptr < env + env_len && *ptr == 'T') {
    size_t rcptlen = strlen(++ptr);
    memcpy(tmp, ptr, rcptlen);
    tmp[rcptlen] = '\n';
    tmp += rcptlen + 1;
    ptr += rcptlen + 1;
    ++count;
  }
  *tmp = 0;
  mysetenv("QMAILRCPTS", buf, tmp-buf);
  mysetenvu("NUMRCPTS", count);
  free(buf);
}

void parse_envelope(void)
{
  const char* env;
  size_t offset;
  if ((env = mmap(0, env_len, PROT_READ, MAP_PRIVATE, ENVIN, 0)) == MAP_FAILED)
    exit(QQ_OOM);
  offset = parse_sender(env);
  parse_rcpts(env, offset);
  munmap((void*)env, env_len);
}

/* Create a temporary invisible file opened for read/write */
int mktmpfile()
{
  char filename[sizeof(TMPDIR)+19] = TMPDIR "/fixheaders.XXXXXX";
  
  int fd = mkstemp(filename);
  if(fd == -1)
    exit(QQ_WRITE_ERROR);

  /* The following makes the temporary file disappear immediately on
     program exit. */
  if(unlink(filename) == -1)
    exit(QQ_WRITE_ERROR);
  
  return fd;
}

/* Renumber from one FD to another */
void move_fd(int currfd, int newfd)
{
  if (currfd == newfd)
    return;
  if (dup2(currfd, newfd) != newfd)
    exit(QQ_WRITE_ERROR);
  if (close(currfd) != 0)
    exit(QQ_WRITE_ERROR);
}

/* Copy from one FD to a temporary FD */
void copy_fd(int fdin, int fdout, size_t* var)
{
  unsigned long bytes;
  int tmp = mktmpfile();
  
  /* Copy the message into the temporary file */
  for(bytes = 0;;) {
    char buf[BUFSIZE];
    ssize_t rd = read(fdin, buf, BUFSIZE);
    if(rd == -1)
      exit(QQ_WRITE_ERROR);
    if(rd == 0)
      break;
    if(write(tmp, buf, rd) != rd)
      exit(QQ_WRITE_ERROR);
    bytes += rd;
  }

  close(fdin);
  if (lseek(tmp, 0, SEEK_SET) != 0)
    exit(QQ_WRITE_ERROR);
  move_fd(tmp, fdout);
  *var = bytes;
}

struct command
{
  char** argv;
  struct command* next;
};
typedef struct command command;

/* Split up the command line into a linked list of seperate commands */
command* parse_args(int argc, char* argv[])
{
  command* tail = 0;
  command* head = 0;
  while(argc > 0) {
    command* cmd;
    int end = 0;
    while(end < argc && strcmp(argv[end], "--"))
      ++end;
    if(end == 0)
      exit(QQ_INTERNAL);
    argv[end] = 0;
    cmd = malloc(sizeof(command));
    cmd->argv = argv;
    cmd->next = 0;
    if(tail)
      tail->next = cmd;
    else
      head = cmd;
    tail = cmd;
    ++end;
    argv += end;
    argc -= end;
  }
  return head;
}

static void mktmpfd(int fd)
{
  int tmp;
  close(fd);
  tmp = mktmpfile();
  move_fd(tmp, fd);
}

static void move_unless_empty(int src, int dst, const void* reopen,
			      size_t* var)
{
  struct stat st;
  if (fstat(src, &st) != 0)
    exit(QQ_INTERNAL);
  if (st.st_size > 0) {
    move_fd(src, dst);
    *var = st.st_size;
    if (reopen) {
      mktmpfd(src);
      if (src == ENVOUT)
	parse_envelope();
    }
  }
  else
    if (!reopen)
      close(src);
  if (lseek(dst, 0, SEEK_SET) != 0)
    exit(QQ_WRITE_ERROR);
}

static void read_qqfd(void)
{
  struct stat st;
  char* buf;
  if (fstat(QQFD, &st) != 0)
    exit(QQ_INTERNAL);
  if (st.st_size > 0) {
    if ((buf = malloc(st.st_size + 1)) == 0)
      exit(QQ_INTERNAL);
    if (read(QQFD, buf, st.st_size) != st.st_size)
      exit(QQ_INTERNAL);
    buf[st.st_size] = 0;
    qqargv[0] = buf;
  }
  close(QQFD);
}

/* Run each of the filters in sequence */
void run_filters(const command* first)
{
  const command* c;
  
  mktmpfd(MSGOUT);
  mktmpfd(ENVOUT);

  for(c = first; c; c = c->next) {
    pid_t pid;
    int status;

    mysetenvu("ENVSIZE", env_len);
    mysetenvu("MSGSIZE", msg_len);
    pid = fork();
    if(pid == -1)
      exit(QQ_OOM);
    if(pid == 0) {
      execvp(c->argv[0], c->argv);
      exit(QQ_INTERNAL);
    }
    if(waitpid(pid, &status, WUNTRACED) == -1)
      exit(QQ_INTERNAL);
    if(!WIFEXITED(status))
      exit(QQ_INTERNAL);
    if(WEXITSTATUS(status))
      exit((WEXITSTATUS(status) == QQ_DROP_MSG) ? 0 : WEXITSTATUS(status));
    move_unless_empty(MSGOUT, MSGIN, c->next, &msg_len);
    move_unless_empty(ENVOUT, ENVIN, c->next, &env_len);
    if (lseek(QQFD, 0, SEEK_SET) != 0)
      exit(QQ_WRITE_ERROR);
  }
}

int main(int argc, char* argv[])
{
  const command* filters;
  
  filters = parse_args(argc-1, argv+1);

  if ((qqargv[0] = getenv("QQF_QMAILQUEUE")) == 0)
    qqargv[0] = QMAIL_QUEUE;

  mysetenvu("QMAILPPID", getppid());

  copy_fd(0, 0, &msg_len);
  copy_fd(1, ENVIN, &env_len);
  parse_envelope();
  mktmpfd(QQFD);

  run_filters(filters);

  read_qqfd();
  move_fd(ENVIN, 1);
  execv(qqargv[0], (char**)qqargv);
  return QQ_INTERNAL;
}
